test/functional/tests/upgrade_in_flight/upgrage_test_utils.py

```diff
@@ -0,0 +1,285 @@
+            )
+        ).config_stats
+    except CmdException:
+        TestRun.LOGGER.error(
+            f"Cache {int(original.cache_id)} in {original.write_policy} on {original.cache_dev} "
+            f"is not running after upgrade"
+        )
+        return
+
+    error_message_template = f"Cache {int(original.cache_id)} " + "expected {}, got {}"
+    stats_to_compare = [
+        "cache_dev",
+        "eviction_policy",
+        "cleaning_policy",
+        "promotion_policy",
+        "cache_line_size",
+        "status",
+        "write_policy",
+    ]
+
+    __compare_stats(original, actual, stats_to_compare, error_message_template)
+
+
+def upgrade_compare_core_conf_section(original, cache_id):
+    cache_id = int(cache_id)
+    try:
+        actual = CoreStats(
+            get_statistics(
+                cache_id=cache_id,
+                core_id=int(original.core_id),
+                filter=[casadm.StatsFilter.conf],
+            )
+        ).config_stats
+    except CmdException:
+        TestRun.LOGGER.error(
+            f"Cache {cache_id} has missing core {original.core_id} on {original.core_dev}"
+        )
+        return
+
+    error_message_template = (
+        f"Cache {cache_id}, core {int(original.core_id)} " + "expected {}, got {}"
+    )
+    stats_to_compare = [
+        "exp_obj",
+        "core_dev",
+        "status",
+        "seq_cutoff_threshold",
+        "seq_cutoff_policy",
+    ]
+
+    __compare_stats(original, actual, stats_to_compare, error_message_template)
+
+
+def __compare_stats(expected, actual, attributes_list, error_message_template):
+    for stat in attributes_list:
+        original_val = getattr(expected, stat)
+        actual_val = getattr(actual, stat)
+
+        if original_val != actual_val:
+            TestRun.LOGGER.error(
+                error_message_template.format(original_val, actual_val)
+            )
+
+
+def upgrade_verify_version_cmd(
+    original_cas_disk_version, original_cas_cache_version, original_casadm_version
+):
+    final_casadm_version = get_casadm_version()
+    final_cas_cache_version = get_cas_cache_version()
+    final_cas_disk_version = get_cas_disk_version()
+
+    if original_casadm_version != final_casadm_version:
+        TestRun.LOGGER.error(
+            f"casadm {final_casadm_version} version is installed "
+            f"instead of expected {original_casadm_version}"
+        )
+
+    if original_cas_cache_version != final_cas_cache_version:
+        TestRun.LOGGER.error(
+            f"cas_cache.ko {final_cas_cache_version} version is installed "
+            f"instead of expected {original_cas_cache_version}"
+        )
+
+    if original_cas_disk_version != final_cas_disk_version:
+        TestRun.LOGGER.error(
+            f"cas_disk.ko {final_cas_disk_version} version is installed "
+            f"instead of expected {original_cas_disk_version}"
+        )
+
+
+def upgrade_prepare_caches(cache_line_size):
```

Comment:
I don't know if You can use `TestRun.step` inside function called in tests. As I know we should avoid that.